File: backend/src/main/java/com/rungo/api/domain/registration/repository/RegistrationCancelHistoryRepository.java
package com.rungo.api.domain.registration.repository;

import com.rungo.api.domain.registration.entity.RegistrationCancelHistory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

public interface RegistrationCancelHistoryRepository extends JpaRepository<RegistrationCancelHistory, Long> {
    Page<RegistrationCancelHistory> findByUserId(Long userId, Pageable pageable);
}

File: backend/src/main/java/com/rungo/api/domain/registration/repository/RegistrationCancelHistoryRepository.kt
package com.rungo.api.domain.registration.repository

import com.rungo.api.domain.registration.entity.RegistrationCancelHistory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param

interface RegistrationCancelHistoryRepository : JpaRepository<RegistrationCancelHistory, Long> {
    fun findByUserId(userId: Long, pageable: Pageable): Page<RegistrationCancelHistory>

    @Modifying(clearAutomatically = true, flushAutomatically = true)
    @Query("DELETE FROM RegistrationCancelHistory h WHERE h.marathonId IN :marathonIds")
    fun deleteAllByMarathonIdIn(@Param("marathonIds") marathonIds: List<Long>)
}
